### PR TITLE
Adds WPAndroid Integration document

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR adds documentation about how `react-native-bridge` is integrated to WPAndroid with the new composite & binary build flows. It also includes instructions about how to achieve specific flows.

### CODE REVIEW

File: WPAndroid Integration.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+### Summary
+
+WPAndroid by default integrates `react-native-bridge` as a binary dependency that's fetched from a remote maven repo. The CI in this repo will deploy a new version each time a commit is pushed to `develop` or an open PR, as well as each time a tag is created. This is the only required dependency. Whenever a new version is deployed, we can simply update `ext.gutenbergMobileVersion` in [build.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/build.gradle).
```
Comment (mokagio) [inline WPAndroid Integration.md]: Totally out of scope, but I thought I'd mention it here. 

For a while now, I've been writing Markdown using a simplified version of [_Semantic Line Breaks_](https://sembr.org/): add a line break each ~~substantial unit of thought~~ sentence.

```suggestion
WPAndroid by default integrates `react-native-bridge` as a binary dependency that's fetched from a remote maven repo.
The CI in this repo will deploy a new version each time a commit is pushed to `develop` or an open PR, as well as each time a tag is created.
This is the only required dependency. Whenever a new version is deployed, we can simply update `ext.gutenbergMobileVersion` in [build.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/build.gradle).
```

From a code review and source code management point of view, semantic line breaks make it easier to suggest and identify changes, because they reduce the scope from the paragraph to the line.

File: WPAndroid Integration.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+- [How to](#how-to)
+    - [How to test a gutenberg-mobile PR in WPAndroid](how-to-test-a-gutenberg-mobile-pr-in-wpandroid)
+    - [How to share WIP WPAndroid APK](how-to-share-wip-wpandroid-apk)
+    - [How to checkout gutenberg-mobile to deployed binary version](#how-to-checkout-gutenberg-mobile-to-deployed-binary-version)
+
+---
+
+### Work with local gutenberg-mobile checkout
+
+- Make sure you run `npm install` or `npm ci`
```
Comment (mokagio) [inline WPAndroid Integration.md:30]: I think it would be clearer to suggest only one option here.

I think `npm ci` would be best here, because we're interested in _consuming_ the local checkout more than modifying its libraries versions (something that could happen with `npm install`).

File: WPAndroid Integration.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+- Copy [local-builds.gradle-example](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/local-builds.gradle-example) renaming it to `local-builds.gradle`
+- Update `localGutenbergMobilePath`
```
Comment (mokagio) [inline WPAndroid Integration.md]: Even though it may seem obvious, I'd add:

```suggestion
- Update `localGutenbergMobilePath` with the path to gutenberg-mobile in your machine
```

File: WPAndroid Integration.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+- Run metro server with `npm run start:reset`
+
+**How it works:**
```
Comment (mokagio) [inline WPAndroid Integration.md]: Nitpick, since this is not a standalone sentence:

```suggestion
#### How it works
```

File: WPAndroid Integration.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+When [settings.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/settings.gradle) finds `localGutenbergMobilePath` in `local-builds.gradle`, it'll substitute the below binary dependency with the local folder:
+
+```
+implementation "$rootProject.gradle.ext.gutenbergMobileBinaryPath:$rootProject.ext.gutenbergMobileVersion"
+```
+
+`react-native-bridge` project is setup to include all the other projects, so no other action is necessary.
```
Comment (mokagio) [inline WPAndroid Integration.md]: Nitpick

```suggestion
The `react-native-bridge` project is setup to include all the other projects, so no other action is necessary.
```

or

```suggestion
`react-native-bridge` is setup to include all the other projects, so no other action is necessary.
```

The rationale, possibly flawed, behind my suggestion is that "`react-native-bridge`" is the "first name" of the of the sentence's subject while "project" is a specification of what `react-native-bridge` is (_I think..._)

For example, "Gio" is my name and I'm a reviewer of this PR. "Gio reviewer makes nitpicky comments" doesn't sound as right as "Gio makes nitpicky comments" or "The reviewer Gio makes nitpicky comments". 

**This is an interesting discussion on grammar/writing but either suggestion provides a small epsilon improvement to the document, so 🤷‍♂️**

File: WPAndroid Integration.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+---
+
+### Deploy a new version of `react-native-bridge` from CI
+
+There are a few different ways to do this:
+
+1. Open a new PR or push a new commit to an open PR which will be published as `<PR number>-<commit full SHA1>`
+2. Merge a PR to `develop` which will be published as `<PR number>-<commit full SHA1>`
```
Comment (mokagio) [inline WPAndroid Integration.md]: Probably a copy-paste slip. I thought we agreed to use:

```suggestion
2. Merge a PR to `develop` which will be published as `develop-<commit full SHA1>`
```

From a technical point of view, I don't think CircleCI can obtain the PR identifier of the commit that triggered the build on `develop`.

File: WPAndroid Integration.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+3. Create a new tag which will be published as `<tag name>`
+
+**How it works:**
+
+CI will run the following commands:
+
+- `npm install`
```
Comment (mokagio) [inline WPAndroid Integration.md]:
```suggestion
- `npm ci`
```

Reference [this](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3230/files#r591378300) and [this](https://github.com/wordpress-mobile/gutenberg-mobile/compare/mokagio/publish-to-bintray-with-script...mokagio/publish-to-bintray-with-script-demo-2#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R317).

File: WPAndroid Integration.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+- `npm run bundle:android`
+- `gutenberg/packages/react-native-bridge/android/publish-aztec-and-bridge.sh <version>`
```
Comment (mokagio) [inline WPAndroid Integration.md:65]:
```suggestion
- `gutenberg/packages/react-native-bridge/android/publish-aztec-and-bridge.sh <version>`, where `version` is a value computed in CI following the rules described above
```

Or do you think it's too verbose?

File: WPAndroid Integration.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+---
+
+### Manually deploy a new version of `react-native-bridge`
+
+_**IMPORTANT:** This should only be used for testing and a version deployed as such shouldn't be merged to WPAndroid's `develop`. This is because we'd bypass the guarantees CI would otherwise give us._
```
Comment (mokagio) [inline WPAndroid Integration.md]: I'm not sure I understand this note.

The following suggestion is the combination of my understanding of the process and of the sentences themselves.

```suggestion
_**IMPORTANT:** This should only be used for testing a version not yet deployed and as such the new version value shouldn't be committed to WPAndroid's `develop`. This is because the process bypasses the guarantees CI would otherwise give us._
```

File: WPAndroid Integration.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+Assuming that there is no open WPAndroid PR yet:
+
+1. Setup WPAndroid to [work with local gutenberg-mobile checkout](#work-with-local-gutenberg-mobile-checkout)
+2. Switch `gutenberg-mobile` to the branch you're testing
+3. Run the app
+
+---
+
+### How to share WIP WPAndroid APK
```
Comment (mokagio) [inline WPAndroid Integration.md]: Nitpick

```suggestion
### How to share a WIP WPAndroid APK
```

or 

```suggestion
### How to share WIP WPAndroid APKs
```

File: WPAndroid Integration.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
+### Summary
```
Comment (mokagio) [inline WPAndroid Integration.md:3]: I think this could benefit from an `h1` title:

```suggestion
# WordPress Android Integration Guide

### Summary
```

When looking at the doc from GitHub, it's not immediately clear what it is about because the file name is as prominent as the `h3` in the rendered markdown:

image

**Review by mokagio — COMMENTED** [review]
Thank you for writing this up! It will be valuable for existing and future contributors.

Most of my comments are nitpick, with only two sentences requiring clarification.

File: WPAndroid Integration.md (hunk 2 above)
Comment (oguzkocer) [reply WPAndroid Integration.md:30]: Although I agree with you, the reason I wrote it this way is because `npm install` is the default suggestion everywhere in `gutenberg-mobile` and `gutenberg` repos, at least as far as I can tell. Going against that in this document didn't feel right to me.

I'll leave this decision to @mchowning.

File: WPAndroid Integration.md (hunk 3 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: I've opted to the following instead:

> - Update `localGutenbergMobilePath`to your local `gutenberg-mobile`checkout

Simply because I used `local checkout` terminology everywhere else in the document. 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1

File: WPAndroid Integration.md (hunk 4 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

File: WPAndroid Integration.md (hunk 5 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

File: WPAndroid Integration.md (hunk 6 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: Yeap, copy-paste slip. Fixed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

File: WPAndroid Integration.md (hunk 7 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: Fixed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

File: WPAndroid Integration.md (hunk 8 above)
Comment (oguzkocer) [reply WPAndroid Integration.md:65]: Is this really necessary? We already have the main section explaining the version value and `how it works` section is supposed to just give extra insight into what's happening.

File: WPAndroid Integration.md (hunk 9 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: I didn't fully understand your suggestion either. I re-worked this in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1, let me know what you think.

File: WPAndroid Integration.md (hunk 10 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

File: WPAndroid Integration.md (hunk 11 above)
Comment (oguzkocer) [reply WPAndroid Integration.md:3]: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

File: WPAndroid Integration.md (hunk 1 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: TIL. Added line breaks to the whole document in 56cd0065eebe143c6034380851f4f4dbd604f4d8.

**Comment by oguzkocer** [comment]
@mokagio Thanks for the review. I addressed/replied to all your comments. This is ready for another look!

**Comment by AmandaRiu** [comment]
Really nice documentation @oguzkocer!

File: WPAndroid Integration.md (hunk 2 above)
Comment (mchowning) [reply WPAndroid Integration.md:30]: I would either leave it as-is or just have `npm install` since that is the default suggestion elsewhere.

> we're interested in consuming the local checkout more than modifying its libraries versions

Is this necessarily true? Won't some devs be using the local gutenberg-mobile checkouts for development?

File: WPAndroid Integration.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    - [How to checkout gutenberg-mobile to deployed binary version](#how-to-checkout-gutenberg-mobile-to-deployed-binary-version)
+
+---
+
+### Work with local gutenberg-mobile checkout
+
+- Make sure you run `npm install` or `npm ci`
+- Copy [local-builds.gradle-example](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/local-builds.gradle-example) renaming it to `local-builds.gradle`
+- Update `localGutenbergMobilePath`to your local `gutenberg-mobile`checkout
+- Run metro server with `npm run start:reset`
```
Comment (mchowning) [inline WPAndroid Integration.md]: I find these instructions a bit hard to follow because we don't say _where_ to run each command. Some should be run in `mobile-gutenberg/`, but others happen in the WPAndroid root (i.e., the copying and renaming of `local-builds.gradle-example` which doesn't say where the file should be copied to).

File: WPAndroid Integration.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+---
+
+### Manually deploy a new version of `react-native-bridge`
+
+_**IMPORTANT:** This should only be used for testing.
+We shouldn't commit a manually deployed version to WPAndroid's `develop` because the process bypasses the guarantees CI would otherwise give us._
+
+- Make sure you run `npm install` or `npm ci` first
+- Run `npm run bundle:android` so the JS bundle is created
+- Run `publish-aztec-and-bridge.sh` script in `react-native-bridge`, giving it a version name such as `locally-built-by-<your_name>-<commit full SHA1>`
```
Comment (mchowning) [inline WPAndroid Integration.md]: > giving it a version name

Might be good to be explicit about what "it" is and how you give it a version name, i.e. is this something the script prompts you for or an argument to the script?

File: WPAndroid Integration.md [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+1. Setup WPAndroid to [work with local gutenberg-mobile checkout](#work-with-local-gutenberg-mobile-checkout)
+2. Switch `gutenberg-mobile` to the branch you're testing
+3. Run the app
+
+---
+
+### How to share a WIP WPAndroid APK
+
+1. Open a `gutenberg-mobile` PR which will [publish a new version](#deploy-a-new-version-of-react-native-bridge-from-ci) or [manually publish a new version](#manually-deploy-a-new-version-of-react-native-bridge)
```
Comment (mchowning) [inline WPAndroid Integration.md]: I would rephrase this a bit so there's no chance of someone reading this as saying that opening the PR will "publish a new version or manually publish a new version".
```suggestion
1. Either (a) open a `gutenberg-mobile` PR which will [publish a new version](#deploy-a-new-version-of-react-native-bridge-from-ci) or (b) [manually publish a new version](#manually-deploy-a-new-version-of-react-native-bridge)
```

File: WPAndroid Integration.md [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+2. Open a WPAndroid PR [updating the binary version](#update-wpandroids-binary-version)
+3. Peril will publish a comment to the WPAndroid PR with a link to the APK
+
+Alternatively, once the `react-native-bridge` version is published, we can locally create the APK.
+**Note that there is no way to use composite build to create a WPAndroid APK that can be shared. This is because composite build only works with the local metro server.**
```
Comment (mchowning) [inline WPAndroid Integration.md:112]: Good idea to call this out. 👍

File: WPAndroid Integration.md [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+---
+
+### How to checkout gutenberg-mobile to deployed binary version
+
+Whenever the CI publishes `react-native-bridge`, it will either include the full commit hash or the tag in its name.
+So, we can checkout simply checkout the commit has or the tag.
```
Comment (mchowning) [inline WPAndroid Integration.md]:
```suggestion
So, we can checkout simply checkout the commit hash or the tag.
```

File: WPAndroid Integration.md (hunk 2 above)
Comment (oguzkocer) [reply WPAndroid Integration.md:30]: I am going to leave it as is. Since the CI does `npm ci`, I think it's good to have that as an alternative.

> Is this necessarily true? Won't some devs be using the local gutenberg-mobile checkouts for development?

Yes they would, that's the whole idea.

File: WPAndroid Integration.md (hunk 12 above)
Comment (mchowning) [reply WPAndroid Integration.md]: One way to help this be clearer might be to group the `npm` steps together and the `WPAndroid` steps together instead of having `npm` -> `WPAndroid` -> `npm`. Just a thought.

**Review by mchowning — COMMENTED** [review]
This is really great @oguzkocer ! Thank you so much for putting this together. I left a few comments, but they're all only minor points.

The only larger concern I have is whether a lot of this documentation might make more sense in the WPAndroid repo. Pretty much all of this other than the parts about deploying a new `react-native-bridge` binary are specific to WPAndroid, so it feels to me like those parts belong in WPAndroid since mobile-gutenberg is a library that could be used in other apps (in theory at least 🙂 ). wdyt?

File: WPAndroid Integration.md (hunk 4 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: I realized I didn't do this for the other 2 instances, so updated in ba88fd37a9778222b56f2ed7669286cba9150619.

File: WPAndroid Integration.md (hunk 12 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: I've reworked this in 6a5ca7effc502ea798275d8ced92a54755d14942, let me know if that's a bit clearer.

File: WPAndroid Integration.md (hunk 13 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: I've reworked this in 6a5ca7effc502ea798275d8ced92a54755d14942, let me know if that's clearer.

FYI: The script will also return an example usage if it's not called properly.

File: WPAndroid Integration.md (hunk 14 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: Addressed in 6a5ca7effc502ea798275d8ced92a54755d14942.

File: WPAndroid Integration.md (hunk 16 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: 🤦 Fixed in 6a5ca7effc502ea798275d8ced92a54755d14942.

File: WPAndroid Integration.md (hunk 13 above)
Comment (mchowning) [reply WPAndroid Integration.md]: Looks great! 👍

**Comment by oguzkocer** [comment]
Thanks for the review and the feedback @mchowning! I've addressed them all in 6a5ca7effc502ea798275d8ced92a54755d14942.

> The only larger concern I have is whether a lot of this documentation might make more sense in the WPAndroid repo. Pretty much all of this other than the parts about deploying a new react-native-bridge binary are specific to WPAndroid, so it feels to me like those parts belong in WPAndroid since mobile-gutenberg is a library that could be used in other apps (in theory at least 🙂 ). wdyt?

That's exactly why I thought this was the best place for it. There is nothing special about WPAndroid and another client can easily get the same setup working. I think having a document such as this makes that work a lot simpler because not only it shows what's possible, but also it explains how it works. That was one of the things I kept in mind while composing the document.

Does that make sense to you @mchowning?

**Comment by mchowning** [comment]
>  There is nothing special about WPAndroid and another client can easily get the same setup working. I think having a document such as this makes that work a lot simpler because not only it shows what's possible, but also it explains how it works.

That does make sense. Having so many references to WPAndroid gutenberg-mobile still feels kinda wrong to me. I'll think about this some more and see if any better ideas come to mind. I don't consider this a blocker.

**Review by mchowning — APPROVED** [review]
👍   from my perspective!

**Comment by oguzkocer** [comment]
> That does make sense. Having so many references to WPAndroid gutenberg-mobile still feels kinda wrong to me. I'll think about this some more and see if any better ideas come to mind. I don't consider this a blocker.

Please do. I couldn't figure out a better way to do this, but if you have any ideas, I'd be more than open to it.

File: WPAndroid Integration.md (hunk 12 above)
Comment (oguzkocer) [reply WPAndroid Integration.md]: > One way to help this be clearer might be to group the npm steps together and the WPAndroid steps together instead of having npm -> WPAndroid -> npm. Just a thought.

I actually missed this comment while working on your feedback. So, I am glad I arrived at the same approach :)

File: WPAndroid Integration.md (hunk 2 above)
Comment (mokagio) [reply WPAndroid Integration.md:30]: Cool 👍 

> > Is this necessarily true? Won't some devs be using the local gutenberg-mobile checkouts for development?
>
> Yes they would, that's the whole idea.

Out of curiosity, would that entail much work around the Node modules?

File: WPAndroid Integration.md (hunk 3 above)
Comment (mokagio) [reply WPAndroid Integration.md]: 👍

File: WPAndroid Integration.md (hunk 8 above)
Comment (mokagio) [reply WPAndroid Integration.md:65]: Fair enough. Too verbose 👍

**Review by mokagio — APPROVED** [review]
Thanks for taking my suggestions into consideration @oguzkocer.

✅ 

Regarding the many WPAndroid mentions in a guide that lives outside of that project (see [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3262#issuecomment-800633091)) I think we could take a second look at this once the dust settles and reword it a bit to make it clear that this is an integration guide form `gutenberg-mobile` in Android. WordPress Android is but one possible `gutenberg-mobile` client, which we're using here as an example since it's the only current client.

That doesn't have to happen soon and it's kind of "premature optimization" until there'll be a genuine need for a new client app.

Waiting to update this documentation will also give us time to learn from using the setup in the real world.

**Comment by mokagio** [comment]
I'm going to hold merging this into the CI deployment branch for a day or two, while I finish working on that part of the code. Just in case I might end up scratching the branch and opening a PR with a new one.

File: WPAndroid Integration.md (hunk 2 above)
Comment (mchowning) [reply WPAndroid Integration.md:30]: > Out of curiosity, would that entail much work around the Node modules?

Not the vast majority of the time, but it could.

**Comment by mokagio** [comment]
> I'm going to hold merging this into the CI deployment branch for a day or two...

I shifted a few things around and was able to focus entirely on this work instead, putting https://github.com/wordpress-mobile/gutenberg-mobile/pull/3230 into a ready to re-review state.

Since @mchowning already approved this, I'm going to merge it now.
